Makefile

```diff
@@ -178,6 +178,15 @@ $(BIN)/$(PROTOC_VERSION_BIN): | $(BIN)
 # Codegen targets
 # ====================================
 
+# IDL submodule must be populated, or files will not exist -> prerequisites will be wrong -> build will fail.
+# Because it must exist before the makefile is parsed, this cannot be done automatically as part of a build.
+# Instead: call this func in targets that require the submodule to exist, so that target will not be built.
```

Comment:
> "this cannot be done automatically as part of a build"

Just curious: why this can be done for thrift but not proto?

Reply:
neither works / it's used in both.  but they are a bit different:
- thrift runs a "read file -> produce generated code" _per thrift file_, and they're passed in as arguments
- protoc runs a _single_ generator call, and it discovers all relevant files in the folders you tell it about

So $(BUILD)/thrift depends on [each thrift file's generation], which means those have to be defined at parse-time, or the makefile won't set up the commands / infer those dependencies (since there are none).  This is difficult to deal with in make, since that "what will be executed" tree is created once at parse-time, and you can't update it later.

Protoc doesn't actually depend on the files at a makefile-level, so as long as the submodule exists before it runs, it's fine.  Or it can easily be made to be fine, anyway, since the dependency tree doesn't need to change whether those files exist or not.

Reply:
ah, and just to be clear: the only reason I'm using `THRIFT_FILES` is because it's easy and sufficient - it's empty if the idls submodule is empty.  `PROTO_FILES` is not, because some of the files are in the `proto` folder instead.